lib/puppet/provider/package/windows/exe_package.rb

```diff
@@ -55,7 +57,28 @@ def match?(resource)
     end
 
     def self.install_command(resource)
-      munge(resource[:source])
+      file_location = resource[:source]
+      if file_location.start_with?('http://', 'https://')
+        begin
+          uri = URI(Puppet::Util.uri_encode(resource[:source]))
+        rescue => detail
+          raise Puppet::Error.new(_("Error when installing %{package}, parsing URI: %{detail}") % { package: resource[:name] ,detail: detail.message}, detail)
+        end
+
+        file_location = File.join(Dir.tmpdir, File.basename(resource[:source]))
```

Comment:
Are the files here cleaned by the provider? What happens if there is a name collision from multiple exe package resources?

Reply:
It's possible for providers to implement `post_resource_eval` and perform some cleanup, like the [`posix` file provider](https://github.com/puppetlabs/puppet/blob/2f14ddd269c26cba36771d9b3464c2c29422e46a/lib/puppet/provider/file/posix.rb#L13-L15)

Reply:
Since `post_resource_eval` is a provider class method, you could define a class method to register tmpfiles, eg
  
  ```ruby
    def self.register(path)
      @paths ||= []
      @paths << path
    end
  ```
  
  Then call that method after creating the tempfile:
  
  ```ruby
  self.class.register(file_location)
  ```
  
  and delete those files in `post_resource_eval`:
  
  ```ruby
  def self.post_resource_eval
    @paths.each { |p| Puppet::FileSystem.unlink(p) } if @paths
  end
  ```